### ISSUE TITLE
Add support for the GSSAPI SASL mechanism

### ISSUE THREAD

**Issue by JMyklebust** [jstedfast/MailKit#1249]
Ref. issue #890, there is a Kerberos .net library here: https://github.com/dotnet/Kerberos.NET.

Would i be possible to get support for kerberos authentication in MailKit with that at some point?

**Comment by jstedfast** [comment]
I actually came across that the other day as I was searching to see if anyone had implemented a managed NTLM library for .NET and thought about possibly adding Kerberos support.

**Comment by SuperFlue** [comment]
Would be great to see Kerberos support, NTLM would be convenient. 
But security wise it would better to have support for Kerberos.

**Comment by jstedfast** [comment]
Just to make sure I know which SASL mechanism you guys want, what are the AUTH mechanism names that your SMTP/POP3/IMAP servers are reporting?

I'm assuming GSSAPI?

Back in the day (20 years ago), KERBEROS_V4 and GSSAPI were the 2 I had implemented in C, but these days it looks like there might be more variants than that:

https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml

**Comment by JMyklebust** [comment]
In my specific case it's a SMTP server that reports GSSAPI yes.

**Comment by jstedfast** [comment]
This doesn't look like it will be as easy as I had hoped. Kerberos.NET doesn't implement any of the GSS API's. Kerberos is only a small part of GSS afaict.

**Comment by jstedfast** [comment]
https://datatracker.ietf.org/doc/html/rfc4752

**Comment by JMyklebust** [comment]
That's some big specifications.... 😅 

Maybe a somewhat naive question: 
Could you leverage the Windows APIs for this?
Example: https://docs.microsoft.com/en-us/windows/win32/secauthn/sspi-kerberos-interoperability-with-gssapi
Obviously that wont be a solid cross-platform solution, but as a stopgap?

Of course, if that just adds lots of extra work that you cannot really re-use for cross-platform. Then I guess I'll just sit tight til you hopefully find time to implement this in a proper way.

MailKit is a great library regardless, keep up the good work. 😄

**Comment by Evengard** [comment]
Erhm, I had success with GSSAPI authentication on Ldap (https://github.com/dsbenghe/Novell.Directory.Ldap.NETStandard) with Kerberos.NET, aka fully managed. What is not supported in Kerberos.NET that you need? It have an EncodeGssApi function which should be enough. Or is there smth else?

**Comment by jstedfast** [comment]
Need a bit more than an encoder.

Btw, your ldap link does not implement GSS-API either 😁

**Comment by Evengard** [comment]
Yeah, because I extended it for the said implementation. Work-related code. There was nothing special though. Basically after reading https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-SSEAN/%5bMS-SSEAN%5d.pdf and https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-SPNG/%5bMS-SPNG%5d.pdf, and seeing that all theese token implementations are present in Kerberos.NET, shouldn't be too complex to implement.

**Comment by jstedfast** [comment]
> shouldn't be too complex to implement

Good, then I'll expect that patch from you by tomorrow ;-)

**Comment by Evengard** [comment]
xD If only I had more time that could very well happen xD

**Comment by jstedfast** [comment]
I'm thinking that this should probably be a separate MailKit.Security.Gssapi nuget package if/when I (or someone else) ever get around to implementing it.

I'd love to see this supported in MailKit, but I'm just not seeing the demand so it's been a very low priority for me (especially now that I have had so little free time to work on this and other side projects).

**Comment by jstedfast** [comment]
https://github.com/dotnet/Kerberos.NET/issues/9

**Comment by jstedfast** [comment]
This is in the works and should be available "soon".

**Comment by jstedfast** [comment]
Okay, fix is in - but I need testers.

**Comment by JMyklebust** [comment]
Oh! That's pretty nice!
I will try to get around to test this on my side, since I requested it after all. :)
Our environment has had a few changes but the use case should still be valid here.

Just to verify though, it's not in a release and I should test a build from source?

**Comment by jstedfast** [comment]
You can use the myget packages linked in the README (table at the top)

**Comment by JMyklebust** [comment]
Can confirm it works as intended. Tested sending from PowerShell 7 via an Exchange server and authenticating with Gssapi.

Here is my POC script for others to easily test too.
````powershell
# POC script for using MailKit to send mail in PowerShell 7
# https://github.com/jstedfast/MailKit?tab=readme-ov-file
#Requires -Version 7.4

# Download the nuget packages and extract them somewhere
# Then adjust the imports accordingly

# https://www.nuget.org/packages/BouncyCastle.Cryptography
Import-Module "R:\bouncycastle.cryptography.2.5.1\lib\net6.0\BouncyCastle.Cryptography.dll"
# https://www.myget.org/feed/mimekit/package/nuget/MimeKit
Import-Module "R:\MimeKit.4.11.0.1569\lib\net8.0\MimeKit.dll"
# https://www.myget.org/feed/mimekit/package/nuget/MailKit
Import-Module "R:\MailKit.4.11.0.1357\lib\net8.0\MailKit.dll"

# The SMTP server to send via
$Server = ""
# The port to use, in case of Exchange server we use the SMTP Submission port 587
$Port = 587
# The mail address we send from
$From = ""
# The mail address we send to
$To = ""

# Build our message
$Message = [MimeKit.MimeMessage]::new()
$Message.From.Add([MimeKit.MailboxAddress]::new($From,$From))
$Message.To.Add([MimeKit.MailboxAddress]::New($To,$To))
$Message.Subject = "Hello world from MailKit"
$Message.Body = [MimeKit.TextPart]::new("plain")
$Message.Body.Text = "This is a test"

# Set up SMTP client and send the message
$SmtpClient = [MailKit.Net.Smtp.SmtpClient]::new()
$SmtpClient.Connect($Server,$Port,$false)
# Authenticate using GSSAPI
$SmtpClient.Authenticate([MailKit.Security.SaslMechanismGssapi]::new())
$SmtpClient.Send($Message)
$SmtpClient.Disconnect($true)
````

**Comment by jstedfast** [comment]
MailKit 4.12.0 has been released with this feature.

**Comment by ricardoboss** [comment]
Maybe I am doing something wrong, but this doesn't work for me.

Using this code:

```csharp
using var client = new ImapClient();
await client.ConnectAsync("imap.my.mail.host", port: 993, SecureSocketOptions.SslOnConnect, cancellationToken);
var credentials = new SaslMechanismGssapi("info@my.mail.host", "some-secure-password");
await client.AuthenticateAsync(credentials, cancellationToken);
```

it throws at the `new SaslMechanismGssapi(...)` line with the following exception:

```
System.NotSupportedException: The requested security package is not supported.
   at System.Net.SSPIWrapper.GetVerifyPackageInfo(ISSPIInterface secModule, String packageName, Boolean throwIfMissing)
   at System.Net.NegotiateAuthenticationPal.WindowsNegotiateAuthenticationPal..ctor(NegotiateAuthenticationClientOptions clientOptions)
   at System.Net.NegotiateAuthenticationPal.Create(NegotiateAuthenticationClientOptions clientOptions)
```

My guess is I am doing something wrong, but I have next to no knowledge about mail protocols, so I'm not even sure how to debug this. Can someone here chime in, since this is most appropriate issue I could find?

---

Forgot to add some important information:

- `Mailkit` version: `4.13.0`
- .NET SDK version: `9.0.302`

**Comment by jstedfast** [comment]
What OS is this on?

This will only work on Windows, most likely.

**Comment by ricardoboss** [comment]
It was on Windows.

Well, it seems Rider wanted to pause execution because of that exception, but it was actually being handled by MailKit. So that was not the actual issue. After letting the code run, I got this new exception:

```
MailKit.Security.SaslException: GSSAPI authentication error: Consistency checks performed on the credential failed.
   at MailKit.Security.SaslMechanismNegotiateBase.GetChallengeResponse(ReadOnlySpan`1 challenge)
   at MailKit.Security.SaslMechanismNegotiateBase.Challenge(Byte[] token, Int32 startIndex, Int32 length, CancellationToken cancellationToken)
   at MailKit.Security.SaslMechanism.ChallengeAsync(Byte[] token, Int32 startIndex, Int32 length, CancellationToken cancellationToken)
   at MailKit.Security.SaslMechanism.ChallengeAsync(String token, CancellationToken cancellationToken)
   at MailKit.Net.Imap.ImapClient.AuthenticateAsync(SaslMechanism mechanism, CancellationToken cancellationToken)
```

...which finally let me realize: I don't actually want GSSAPI credentials for this authentication flow.

I switched to `SaslMechanismPlain` instead and now it's working. (not sure if this is less secure or what alternatives there are... but it works)

**Comment by jstedfast** [comment]
GSSAPI is a Kerberos-based authentication mechanism that typically uses a cached "Single Signon" token to authenticate and doesn't typically use a "username" and "password", I don't think.

I wouldn't expect that GSSAPI would work for you *unless* you are part of an Enterprise that has it configured for you.

> I switched to SaslMechanismPlain instead and now it's working. (not sure if this is less secure or what alternatives there are... but it works)

It's *definitely* less secure, but unless you have Kerberos set up properly on your system & domain, it's not likely you'd be able to use that.
